src/replaying.jl

```diff
@@ -61,3 +61,36 @@ function avoid_allocations(record, f, args...; kwargs...)
     ctx = new_replay_ctx(record)
     return Cassette.overdub(ctx, f, args...; kwargs...)
 end
+
+struct FrozenFunction{F}
+    f::F
+    ctx::Dict{Tuple, ReplayCtx}
+    FrozenFunction(f) = new{typeof(f)}(f, Dict{Tuple, ReplayCtx}())
+end
+
+@generated function (f::FrozenFunction)(xs...)
+    return quote
+        if haskey(f.ctx, $xs)
+            ctx = f.ctx[$xs]
+            # step = ctx.metadata.step::Ref{Int}
+            ctx.metadata.step[] = 1
+            return Cassette.overdub(ctx, f.f, xs...)
+        else
+            x, record = record_allocations(f.f, xs...)
+            ctx = AutoPreallocation.new_replay_ctx(record)
+            f.ctx[$xs] = ctx
+            return x
+        end
+    end
+end
+
+"""
+    freeze(f)
+
+Freeze a function. This will freeze the allocation behaviour of the function by creating
+a [`FrozenFunction`](@ref). This function will record all the allocations at the first run,
+then in the following run, it will not allocate anymore.
+"""
+freeze(f) = FrozenFunction(f)
```

Comment:
Is `freeze` really the clearest word for this?

`freeze` to me means to take a mutable datastructure and generate an immutable datastructure from it.

Reply:
umm, not sure if there is a better one? I feel in the end only one short API from this package is needed. I thought it could mean freeze the allocation behaviour of a function.

Reply:
I think we need a better-one.
Maybe even just `auto_preallocate` ?

Reply:
umm, it's a bit long, isn't it? what about just `preallocate`?

Reply:
I also used to use `hoist` in my own implementation.

Reply:
Yeah, lets go with `preallocate`

Reply:
done.